pyanaconda/modules/subscription/service_observer.py

```diff
@@ -0,0 +1,117 @@
+#
+# Copyright (C) 2020  Red Hat, Inc.  All rights reserved.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+import time
+from threading import RLock, Timer
+
+from pyanaconda.core import util
+from pyanaconda.anaconda_loggers import get_module_logger
+from dasbus.namespace import get_namespace_from_name, get_dbus_path
+from dasbus.client.observer import DBusObserver, DBusObserverError
+
+log = get_module_logger(__name__)
+
+# How long to wait for a systemd unit backed DBus service
+# to become available (in seconds). This timeout is based
+# on the default 90 second timeout for manual systemd unit
+# activation, which is also 90 seconds.
+SERVICE_ACTIVATION_TIMEOUT = 90.0
+# how often to check if service became available ? (in seconds)
+SERVICE_CHECK_INTERVAL = 0.1
+
+class ServiceObserver(DBusObserver):
+    """Observer of a systemd unit backed DBus service."""
+
+    def __init__(self, message_bus, dbus_service_name, systemd_unit_name,
+                 timeout=SERVICE_ACTIVATION_TIMEOUT):
+        """Creates a service observer.
+
+        The service observer will try to start the underlying systemd
+        service and then waits for the DBus API to become available.
+
+        :param message_bus: a message bus
+        :param dbus_service_name: a DBus service name
+        :param systemd_unit_name: a systemd unit name corresponding to the DBus service
+        :param float timeout: how long to wait for the DBus service to start in seconds
+        """
+        super().__init__(message_bus, dbus_service_name)
+        self._message_bus = message_bus
+        self._dbus_service_name = dbus_service_name
+        self._systemd_unit_name = systemd_unit_name
+        self._timeout = timeout
+
+        namespace = get_namespace_from_name(dbus_service_name)
+        self._object_path = get_dbus_path(*namespace)
```

Comment:
How do you plan to access other DBus objects that belong to the RHSM service?

Reply:
Good point, I'll need to look more into that as we do use a bunch of them.